File: cpp/Guesser/guesser.cpp
#include "guesser_realization.h"

#include <iostream>
#include <algorithm>

int main()
{
     std::cout << "Input initial animal: ";
     std::string animal;
     std::getline(std::cin, animal);

     Guesser guesser(animal);
     while (true)
     {
          auto response = guesser.Request();
          std::cout << response.second << "?" << std::endl;
          std::cout << "[Y/N]";

          std::string answer;
          if (!std::getline(std::cin, answer) || answer.empty())
               break;
          std::transform(answer.begin(), answer.end(), answer.begin(),
               [](auto ch) {return static_cast<unsigned char>(std::tolower(ch));});
          if (answer == "yes" || answer == "y")
          {
               guesser.Yes();
               if ("animal" == response.first)
                    std::cout << "Guessed" << std::endl;
          }
          else if (answer == "no" || answer == "n")
          {
               guesser.No();
               if ("animal" == response.first)
               {
                    std::cout << "Input new animal: ";
                    std::string animal;
                    std::getline(std::cin, animal);

                    std::cout << "Input new definition: ";
                    std::string question;
                    std::getline(std::cin, question);

                    guesser.AddDef(question, animal);
               }
          }
          else
               std::cout << "Incorrect input";
     }
}

File: py/guesser.py
"""
Класс для игры "угадай животное".
"""
class Guesser:            
	def __init__(self, initialAnimal):
		"""
		Конструктор. Принимет первое животное, которое знает система.
		"""
		pass

	def Request(self):
		"""
		Метод получения текущего вопроса системы. Может быть вызван несколько раз подряд до получения ответа.
		Кортежи могут быть двух видов:
			1. ("question", <строка с вопросом, как он был введён в систему>);
			2. ("animal", <строка с животным, как оно было введено в систему>).
		"""
		pass

	def Yes(self):
		"""
		На текущий вопрос системы ответ "Да".
		Если Request() возвращает кортеж с "animal", то обновляет сессию.
		"""
		pass

	def No(self):
		"""
		На текущий вопрос системы ответ "Нет".
		"""
		pass

	def AddDef(self, question, animal):
		"""
		Ввести определение, если система не угадала животное.
		Принимает вопрос и животное. Животное является ответом "Да" на вопрос, а ответом "Нет" является животное, получаемое Request().
		Обновляет сессию.
		"""
		pass


if __name__ == "__main__":
	guesser = Guesser(input("Input initial animal: "))
	while True:
		(type, question) = guesser.Request()
		print(question + "?")
		answer = input("[Y/N]")
		answer = answer.lower()
		if answer == "y" or answer == "yes":
			guesser.Yes()
			if (type == "animal"):
				print("Guessed")
		elif answer == "n" or answer == "no":
			guesser.No()
			if (type == "animal"):
				animal = input("Input new animal: ")
				question = input("Input new definition: ")
				guesser.AddDef(question, animal)
		elif answer == "":
			break
		else:
			print("Incorrect input")
